### PR TITLE
feat: added simple claim view.

### PR DESCRIPTION
**Pull request by zhyd1997**
Close: #95

### CODE REVIEW

**Comment by zhyd1997** [comment]
The functionality has already finished and waiting for the lint errors PR is merged.

File: components/cards/ActionForm.tsx [hunk 1]
```diff
@@ -268,6 +275,29 @@ export function ActionForm(props: ActionFormProps) {
...
+                }
+                streamBuffer={
+                  annualNetworkFeeRate
+                    ? truncateEth(
+                        ethers.utils.formatEther(
+                          annualNetworkFeeRate.mul(2 * 60 * 60)
+                        )
+                      )
+                    : "0"
+                }
```
Comment (zhyd1997) [inline components/cards/ActionForm.tsx:297]: @gravenp 

the `Stream` and `StreamBuffer` are calculated here.

File: components/cards/ActionForm.tsx [hunk 2]
```diff
@@ -265,6 +272,25 @@ export function ActionForm(props: ActionFormProps) {
...
+                streamBuffer={
+                  networkFeeRatePerSecond
+                    ? truncateEth(
+                        formatBalance(networkFeeRatePerSecond.mul(2 * 60 * 60)),
+                        18
+                      )
+                    : "0"
+                }
```
Comment (zhyd1997) [inline components/cards/ActionForm.tsx:290]: **Update**: the `Stream` and `StreamBuffer` are calculated here.

Comment (codynhat) [reply components/cards/ActionForm.tsx:290]: Is the buffer here 2 hours? Will it be 2 hours on testnet and 8 hours on a mainnet? If so, can we make this dynamic based on the network?

Comment (codynhat) [reply components/cards/ActionForm.tsx:290]: The constants [here](https://github.com/superfluid-finance/protocol-monorepo/blob/dev/packages/sdk-core/src/constants.ts) may be useful

**Review by codynhat — CHANGES_REQUESTED** [review]

**Comment by codynhat** [comment]
Also see https://github.com/Geo-Web-Project/cadastre/issues/95#issuecomment-1126741952. This PR will need to include the actual claim functionality, too. See the `ClaimAction` component

**Comment by zhyd1997** [comment]
I found here is a bug maybe: https://github.com/Geo-Web-Project/cadastre/blob/5675cce5b77a9618cfc1b9dbf48b73ae65211838/lib/stream-managers/BasicProfileStreamManager.js#L30

both `dataStore` and `didNFT` are `undefined` or empty string `''`, so it caused the error below when claiming.

streamManagerUndefinedError

**Comment by zhyd1997** [comment]
> Also see [#95 (comment)](https://github.com/Geo-Web-Project/cadastre/issues/95#issuecomment-1126741952). This PR will need to include the actual claim functionality, too. See the `ClaimAction` component

Which contract should I use, `geoWebFairLaunchClaimerContract` or `geoReClaimerContract`, I tried the former in this PR, is it correct? 👀

File: components/cards/ClaimAction.tsx [hunk 3]
```diff
@@ -13,78 +15,78 @@ export type ClaimActionProps = SidebarProps & {
...
+      account,
+      newContributionRate,
+      // baseCoord.toString(10),
+      path,
+      {
+        from: account,
+        // value: calculateWeiSubtotalField(displayNetworkFeePayment),
+      }
+    );
+    const receipt = await resp?.wait();
```
Comment (zhyd1997) [inline components/cards/ClaimAction.tsx:81]: Seems that I did something wrong here, but I can not catch it: 👀 

` Error: invalid arrayify value (argument="value", value=, code=INVALID_ARGUMENT, version=bytes/5.4.0)`

Comment (zhyd1997) [reply components/cards/ClaimAction.tsx:81]: It's hard to decide the **3rd** parameters of `claimerContract.claim()`.

Comment (zhyd1997) [reply components/cards/ClaimAction.tsx:81]: Seems that the contract needs 2 parameters, both `path` and `baseCoord`:

https://github.com/Geo-Web-Project/core-contracts/blob/bae960170c9a7d05c5fe3394d81cc1e5967ca661/contracts/FairLaunchClaimer.sol#L52

But how to pass it?

**Comment by gravenp** [comment]
> Which contract should I use, `geoWebFairLaunchClaimerContract` or `geoReClaimerContract`, I tried the former in this PR, is it correct? 👀

Yes,  the `FairLaunchClaimer` is the right one to use here. `Reclaimer` will be used in #102 & #103.

File: components/cards/ClaimAction.tsx (hunk 3 above)
Comment (zhyd1997) [reply components/cards/ClaimAction.tsx:81]: Found typo: `unit64` should be `uint64` when using `abiCoder.encode()`.

Will try it later.

File: components/cards/ClaimAction.tsx [hunk 4]
```diff
@@ -13,78 +15,79 @@ export type ClaimActionProps = SidebarProps & {
...
+      perSecondFeeNumerator,
+      perSecondFeeDenominator
+    );
+    const resp = await claimerContract.claim(
+      account,
+      newContributionRate,
+      ethers.utils.defaultAbiCoder.encode(
+        ["uint64", "uint256[]"],
+        [BigNumber.from(baseCoord.toString(10)), path]
+      ),
```
Comment (zhyd1997) [inline components/cards/ClaimAction.tsx:76]: new error encountered: 
```js
Error: cannot estimate gas; transaction may fail or may require manual gas limit (error=
{"code":3,"data":"...","message":"execution reverted: AccessControl: account 0x6097847fa05eb9af005224eab62784f8419cc454
 is missing role 0xf7db13299c8a9e501861f04c20f69a2444829a36a363cfad4b58864709c75560"}, 
method="estimateGas", transaction={"from":"0x6097847FA05eb9aF005224Eab62784f8419cC454","to":"0xe62D6E28597471a84daa57
392021a21696e17404","data":"...","accessList":null}, code=UNPREDICTABLE_GAS_LIMIT, 
version=providers/5.4.4)
```

I guess I am not the `Admin` according to the contract?

Comment (zhyd1997) [reply components/cards/ClaimAction.tsx:76]: More clear: 

`AccessControl: account ...
 is missing role 0xf7db13299c8a9e501861f04c20f69a2444829a36a363cfad4b58864709c75560`

Comment (zhyd1997) [reply components/cards/ClaimAction.tsx:76]: Do I need to `setClaimer` here or the contract will check `claimer` automatically?

Comment (codynhat) [reply components/cards/ClaimAction.tsx:76]: So the `FairLaunchClaimer` contract should not actually be called directly here, which is why you are seeing a permissions error. See the example [here](https://github.com/Geo-Web-Project/core-contracts/blob/main/docs/AuctionSuperApp.md#claim). A claim is done by creating/updating a flow. If this is the first parcel a user is claiming, the flow should be created like in that example. If there are any existing parcels, the flow should be updated and the rate should be incremented. See the [tests](https://github.com/Geo-Web-Project/core-contracts/blob/main/test/auction-super-app.ts#L208) for an example on how to update an existing flow.

Comment (codynhat) [reply components/cards/ClaimAction.tsx:76]: The baseCoord and path looks correct here. So you just need to change the [claimData](https://github.com/Geo-Web-Project/core-contracts/blob/main/test/auction-super-app.ts#L225) to be the value you have here.

Comment (codynhat) [reply components/cards/ClaimAction.tsx:76]: So, for clarity:
- `claimData` -> baseCoord, path you have here
- `actionData` -> [`forSalePrice`, `claimData`]
- `userData` -> [ Action.CLAIM, `actionData`]

**Comment by zhyd1997** [comment]
The PR still has some problems, figuring it out...

**Comment by zhyd1997** [comment]
I have recalculated `flowRate` and I think this is the last error need to fix:

`cannot estimate gas; transaction may fail or may require manual gas limit`:

gas limit error

File: components/cards/ClaimAction.tsx [hunk 5]
```diff
@@ -13,78 +21,172 @@ export type ClaimActionProps = SidebarProps & {
...
+      .div(SECONDS_IN_YEAR);
+
+    const signer = provider.getSigner() as any;
+
+    if (existingParcels) {
+      // update an exisiting flow
+      console.log("Updting an exisiting flow: ");
+      const existingLicenseID = parcels[0].args[0].toString();
+
+      claimData = ethers.utils.defaultAbiCoder.encode(
```
[the remark this answers is not in the document]
Comment (zhyd1997) [reply components/cards/ClaimAction.tsx:112]: @codynhat do you mean this?

https://github.com/Geo-Web-Project/core-contracts/blob/bb6b1d0e12c03615e1ab21dc7d5e6d19abfb9cc4/test/auction-super-app.ts#L225

[the remark this answers is not in the document]
Comment (codynhat) [reply components/cards/ClaimAction.tsx:112]: Sorry, I was wrong. The example in the tests here is not what you want. Try this:
```
claimData  = ethers.utils.defaultAbiCoder.encode(
        ["uint64", "uint256[]"],
        [BigNumber.from(baseCoord.toString(10)), path]
      )
```

[the remark this answers is not in the document]
Comment (zhyd1997) [reply components/cards/ClaimAction.tsx:112]: I added it here: https://github.com/zhyd1997/cadastre/blob/95abf316f3899c8a188ed03fd2a2494654098bcf/components/cards/ClaimAction.tsx#L151

[the remark this answers is not in the document]
Comment (zhyd1997) [reply components/cards/ClaimAction.tsx:112]: I will specify the logic of claim flow.

File: components/cards/ClaimAction.tsx [hunk 6]
```diff
@@ -13,78 +21,172 @@ export type ClaimActionProps = SidebarProps & {
...
+        receiver: auctionSuperApp.address,
+        superToken: paymentTokenAddress,
+        userData,
+      });
+
+      txn = await sf
+        .batchCall([approveOperation, createFlowOperation])
+        .exec(signer);
+
+      console.log(txn);
```
Comment (zhyd1997) [inline components/cards/ClaimAction.tsx:176]: @codynhat 

Claim flow explanation [1/2]:

> If this is the first parcel a user is claiming, the flow should be created

File: components/cards/ClaimAction.tsx [hunk 7]
```diff
@@ -13,78 +21,172 @@ export type ClaimActionProps = SidebarProps & {
...
+
+    let actionDataToPassInUserData;
+
+    let userData;
+
+    let txn;
+
+    const filter = claimerContract.filters.ParcelClaimed(null, null);
+    const parcels = await claimerContract.queryFilter(filter);
+    const existingParcels = parcels.length > 0;
```
Comment (zhyd1997) [inline components/cards/ClaimAction.tsx:89]: [2/2]

> If there are any existing parcels

File: components/cards/ClaimAction.tsx [hunk 8]
```diff
@@ -13,78 +21,172 @@ export type ClaimActionProps = SidebarProps & {
...
+      });
+
+      const updateFlowOperation = await sf.cfaV1.updateFlow({
+        flowRate: BigNumber.from(existingFlow.flowRate)
+          .add(newFlowRate)
+          .toString(),
+        receiver: auctionSuperApp.address,
+        superToken: paymentTokenAddress,
+        userData,
+      });
```
Comment (zhyd1997) [inline components/cards/ClaimAction.tsx:141]: > the flow should be updated and the rate should be incremented

File: components/cards/ClaimAction.tsx (hunk 6 above)
Comment (codynhat) [reply components/cards/ClaimAction.tsx:176]: Looks like you are missing `actionDataToPassInUserData` in the create flow case.

File: components/cards/ClaimAction.tsx (hunk 7 above)
Comment (codynhat) [reply components/cards/ClaimAction.tsx:89]: Does this check for any existing parcels from any user? You should be able to check if `existingFlow` below is > 0.

File: components/cards/ClaimAction.tsx (hunk 5 above)
Comment (codynhat) [inline components/cards/ClaimAction.tsx:112]: This `claimData` still needs to be updated like the one in the other case.

File: components/cards/ClaimAction.tsx [hunk 9]
```diff
@@ -13,78 +21,172 @@ export type ClaimActionProps = SidebarProps & {
...
+      const existingLicenseID = parcels[0].args[0].toString();
+
+      claimData = ethers.utils.defaultAbiCoder.encode(
+        ["uint256"],
+        [existingLicenseID]
+      );
+
+      actionDataToPassInUserData = ethers.utils.defaultAbiCoder.encode(
+        ["uint256", "bytes"],
+        [displayNewForSalePrice, claimData]
```
Comment (codynhat) [inline components/cards/ClaimAction.tsx:119]: `displayNewForSalePrice` needs to be formatted to wei. So `ethers.utils.parseEther(displayNewForSalePrice)`

File: components/cards/ClaimAction.tsx [hunk 10]
```diff
@@ -13,78 +21,172 @@ export type ClaimActionProps = SidebarProps & {
...
+    const sf = await sfInstance(NETWORK_ID, provider);
+
+    const ethx = await sf.loadSuperToken(paymentTokenAddress);
+
+    const approveOperation = await ethx.approve({
+      receiver: auctionSuperApp.address,
+      amount: ethers.utils.parseEther(displayNewForSalePrice).toString(),
+    });
+
+    const newFlowRate = ethers.utils
```
Comment (codynhat) [inline components/cards/ClaimAction.tsx:101]: The flow rate should be equal to `networkFeeRatePerSecond` in `ActionForm` [here](https://github.com/Geo-Web-Project/cadastre/pull/126/files#diff-958d7daaf87d1b26c5e80206f1032e04b0cb6129727438f9e95eb1c885d63914R110)

File: components/cards/ClaimAction.tsx [hunk 11]
```diff
@@ -13,78 +22,166 @@ export type ClaimActionProps = SidebarProps & {
...
+        .exec(signer);
+
+      console.log(txn);
+    } else {
+      // create a new flow
+      console.log("Creating a new flow: ");
+
+      userData = ethers.utils.defaultAbiCoder.encode(
+        ["uint8", "bytes"],
+        [Action.CLAIM, claimData]
```
Comment (codynhat) [inline components/cards/ClaimAction.tsx:151]: Ok, all this abi.encode stuff is a pain to troubleshoot, but I was able to get a claim to succeed after fixing this: `claimData` -> `actionDataToPassInUserData`.

The claim succeeded, but the event query failed below https://github.com/Geo-Web-Project/cadastre/pull/126/files#diff-77130f94448dec75dfc66a3dc0c3af3f07a3329cf0a300be36bcb042f7f8fd64R176

Comment (zhyd1997) [reply components/cards/ClaimAction.tsx]: @codynhat 

500 error?

subgraph commits 500 error

Comment (codynhat) [reply components/cards/ClaimAction.tsx]: Hmm, I didn't get that. Just an error about `res[0]` being undefined.

Comment (zhyd1997) [reply components/cards/ClaimAction.tsx]: I created a parcel successfully with 500 error, and now I can update the parcel but still get this 500 error:

Screen Shot 2022-05-20 at 12 50 07

**BTW**: could you retrieve it from the geo web subgraph?

Comment (zhyd1997) [reply components/cards/ClaimAction.tsx]: I claimed a new parcel, and I can not get the `res[0] is undefined` error (got an array with 4 elements instead):

Screen Shot 2022-05-20 at 15 54 12

Comment (codynhat) [reply components/cards/ClaimAction.tsx]: Yep, I see some of your claims. I am not able to reproduce the `res[0] is undefined`. Let's monitor to see if it appears again. I am also getting the 500. But let's mark this as complete and address that issue in https://github.com/Geo-Web-Project/cadastre/issues/108

Could you just clean up some of the console.logs? Then I will approve and merge.

Comment (zhyd1997) [reply components/cards/ClaimAction.tsx]: Done

**Review by codynhat — APPROVED** [review]
Looks great! Just some outstanding bugs that will not be fixed in this PR:

- There is currently a bug when writing content to Ceramic
- Parcel info does not load

Both of these to be fixed with https://github.com/Geo-Web-Project/cadastre/issues/96

**Pull request merged**